### PR TITLE
use aws-sdk v3

### PR DESCRIPTION
**Pull request by jsvd**
~~very naive approach to using aws-sdk v3. needs a lot of require cleanup~~

Introduce aws-sdk v3, by including the service-specific gems + aws-sdk-core v3, instead of adding the top level ask-sdk gem that would bring on > 300 dependencies.

This PR also removes the calls to eager loading as in aws-sdk v3 it's no longer necessary since we're loading the gems we need explicitly.

### CODE REVIEW

File: lib/logstash/inputs/s3.rb [hunk 1]
```diff
@@ -99,7 +97,7 @@ def initialize(*params)
   def register
     require "fileutils"
     require "digest/md5"
-    require "aws-sdk-resources"
+    #require "aws-sdk-resources"
```
Comment (jsvd) [inline lib/logstash/inputs/s3.rb:100]:
```suggestion
```

File: logstash-integration-aws.gemspec [hunk 2]
```diff
@@ -19,7 +19,14 @@ Gem::Specification.new do |s|
...
   s.add_runtime_dependency "stud", "~> 0.0.22"
-  s.add_runtime_dependency "aws-sdk", "~> 2"
+  s.add_runtime_dependency "aws-sdk-core", "~> 3"
+  s.add_runtime_dependency "aws-sdk-s3"
+  s.add_runtime_dependency "aws-sdk-sqs"
+  s.add_runtime_dependency "aws-sdk-sns"
+  s.add_runtime_dependency "aws-sdk-cloudwatch"
+  s.add_runtime_dependency "aws-sdk-cloudfront"
+  s.add_runtime_dependency "aws-sdk-resourcegroups"
+#  s.add_runtime_dependency "aws-sdk-resources"
```
Comment (jsvd) [inline logstash-integration-aws.gemspec:29]:
```suggestion
```

File: spec/inputs/s3_spec.rb [hunk 3]
```diff
@@ -4,10 +4,10 @@
 require "logstash/inputs/s3"
 require "logstash/codecs/multiline"
 require "logstash/errors"
-require "aws-sdk-resources"
+#require "aws-sdk-resources"
```
Comment (jsvd) [inline spec/inputs/s3_spec.rb:7]:
```suggestion
```

File: spec/inputs/s3_spec.rb [hunk 4]
```diff
@@ -4,10 +4,10 @@
...
 require "stud/temporary"
 require_relative "../support/helpers"
-require "aws-sdk"
+#require "aws-sdk-core"
```
Comment (jsvd) [inline spec/inputs/s3_spec.rb:10]:
```suggestion
```

File: spec/integration/s3_spec.rb [hunk 5]
```diff
@@ -1,6 +1,6 @@
 require "logstash/devutils/rspec/spec_helper"
 require "logstash/inputs/s3"
-require "aws-sdk"
+require "aws-sdk-core"
```
Comment (jsvd) [inline spec/integration/s3_spec.rb:3]:
```suggestion
```

File: spec/integration/sqs_spec.rb [hunk 6]
```diff
@@ -3,7 +3,7 @@
 require "logstash/inputs/sqs"
 require "logstash/event"
 require "logstash/json"
-require "aws-sdk"
+require "aws-sdk-core"
```
Comment (jsvd) [inline spec/integration/sqs_spec.rb:6]:
```suggestion
```

File: spec/outputs/s3/uploader_spec.rb [hunk 7]
```diff
@@ -2,7 +2,7 @@
 require "logstash/devutils/rspec/spec_helper"
 require "logstash/outputs/s3/uploader"
 require "logstash/outputs/s3/temporary_file"
-require "aws-sdk"
+require "aws-sdk-core"
```
Comment (jsvd) [inline spec/outputs/s3/uploader_spec.rb:5]:
```suggestion
```

File: spec/outputs/s3/write_bucket_permission_validator_spec.rb [hunk 8]
```diff
@@ -1,7 +1,7 @@
 # encoding: utf-8
 require "logstash/devutils/rspec/spec_helper"
 require "logstash/outputs/s3/write_bucket_permission_validator"
-require "aws-sdk"
+require "aws-sdk-core"
```
Comment (jsvd) [inline spec/outputs/s3/write_bucket_permission_validator_spec.rb:4]:
```suggestion
```

File: spec/plugin_mixin/aws_config_spec.rb [hunk 9]
```diff
@@ -1,7 +1,7 @@
 # encoding: utf-8
 require "logstash/devutils/rspec/spec_helper"
 require "logstash/plugin_mixins/aws_config"
-require 'aws-sdk'
+require 'aws-sdk-core'
```
Comment (jsvd) [inline spec/plugin_mixin/aws_config_spec.rb:4]:
```suggestion
```

File: spec/integration/cloudwatch_spec.rb [hunk 10]
```diff
@@ -1,6 +1,6 @@
 require "logstash/devutils/rspec/spec_helper"
 require "logstash/inputs/cloudwatch"
-require "aws-sdk"
+require "aws-sdk-core"
```
Comment (jsvd) [inline spec/integration/cloudwatch_spec.rb:3]:
```suggestion
```

File: spec/inputs/cloudwatch_spec.rb [hunk 11]
```diff
@@ -1,7 +1,7 @@
 require 'logstash/devutils/rspec/spec_helper'
 require 'logstash/devutils/rspec/shared_examples'
 require 'logstash/inputs/cloudwatch'
-require 'aws-sdk'
+require 'aws-sdk-core'
```
Comment (jsvd) [inline spec/inputs/cloudwatch_spec.rb:4]:
```suggestion
```

**Comment by jsvd** [comment]
@robbavey I assigned you as a reviewer for this PR. It targets the [bootstrap_work](https://github.com/logstash-plugins/logstash-integration-aws/tree/bootstrap_work) branch which is a refresh of your #1 + #3. In the `bootstrap_work` branch I tried have each commit be a logical change.
The `use_aws_sdk_v3` branch does the remaining work on moving the integration to sdkv3, by mostly tweaking requires.

**Comment by christianherweg0807** [comment]
Hello @jsvd, I've released a pre-version, for testing purposes.
See: https://rubygems.org/gems/logstash-integration-aws

With this test release, I'll be able to test my own prerelease from logstash-input-s3-sns-sqs.
I´ll overhand the rubygems.org artefakt to you/elastic or delete my gem release.


regards
christian

**Review by robbavey — APPROVED** [review]
These changes LGTM - there are a few changes I'd like to see to the `bootstrap` work branch before we publish, but we can go through those separately to this PR.

**Pull request closed without merging**